### ISSUE TITLE
Tips on troubleshooting/optimising memory usage

### ISSUE THREAD

**Issue by mrmachine** [Suor/django-cacheops#350]
I'm trying to determine why our cacheops instance is storing so much non-volatile data. With `maxmemory` set to 3.5GB and the `volatile-lru` maxmemory policy and LRU cacheops setting, we were getting `-OOM command not allowed when used memory > 'maxmemory'` errors.

We upped our memory limits for now, but analyzing a `dump.rdb` file by loading it into RedisInsight and also loading it into a local Redis server.

RedisInsight says we have 3.25GB of data in sets (3.4M keys) and 972MB in strings (945K keys). Also, 3.31GB of data with no expiry.

When I look at the Memory Analyzer it shows the key with the most memory usage is `conj:TABLENAME:slug=FOO` at 61MB with a size/length of 670K and no expiry.

When I look at the members of this set, it is all `as:HASH` strings. When I attempt to get any of these in redis-cli, I get `(nil)` back.

So I have a few questions that I hope can help me track down which queries and functions are generating the bulk of non-volatile data so we can stay under the `maxmemory` limit.

- Does a the set from above include the `as:` keys for all `@cached_as` functions that include any queryset sample that filters against `TABLENAME__slug=FOO`? So `Model1.objects.filter(slug=FOO)` and `Model2.objects.filter(model1__slug=FOO)` in different `@cached_as` functions?
- Why/how are all the keys in that set that I've tested returning `(nil)`? Are the keys themselves already expired, but the set has continued to accumulate more and more keys over time, never removing expired keys?

**Comment by tumb1er** [comment]
Cache key is computed as md5 checksum of sql query with some other parameters, so if your sql vary - you will get different cache keys. `cached_as` keys and queryset cache keys are also different (imagine they are same and decorated function results are mixed with queryset result cache!)

`conj:` key is a helper set to invalidate all queryset involving lookup on slug=FOO. If you change FOO to MOO, all keys from `conj:` set will be cleaned up and whole cache will remain consistent. Also, `conj:` usually has it's own expire (a bit more than queryset cache keys expire). May be it's some occasional bug, but I guess your never change slugs and they are unique, so your `conj:` is updated very often and never invalidates or expires. In this case it will grow until you notice it) You should monitor all `conj:` keys length to prevent redis OOM, and make a manual script to periodically remove unexistent keys from `conj:` sets

**Comment by mrmachine** [comment]
@tumb1er Thanks. Yes we have a few `Category` objects and their slug NEVER changes, and many things are related to `Category` either directly or through a few FKs.

So we cache many querysets that filter on `...__category__slug=FOO` and functions (via `@cached_as`) functions with queryset samples that filter on `...__category__slug=FOO`.

So if I understand correctly, the `conj:CATEGORY:slug=FOO` key will grow and grow forever, because we never change the Category with `slug=FOO` to have `slug=BAR`.

So this `conj:CATATEGORY:slug=FOO` set will **NEVER** be cleaned up automatically by cacheops?

And when the individual `as:` and `q:` keys containing cache data expire, they are not removed from `conj:` sets that might contain them.

And, are keys in this set for querysets that filter on `...__category__slug=FOO` **in addition** to other filters?

So the `q:` key for `Article.objects.filter(is_active=True, category__slug=FOO)` will go into `conj:CATEGORY:slug=FOO` AND `conj:ARTICLE:is_active=true` sets?

Will the `as:` key for `@cached_as(Article.objects.filter(category__slug=FOO), Other.objects.filter(other_field=BAR)` return value be added to BOTH `conj:CATEGORY:slug=FOO` AND `conj:OTHER:other_field=BAR`?

Could cacheops perhaps treat unique slug fields that never change differently? Or remove expired keys from `conj:` sets periodically?

Or, would we be better served by switching to `volatile-ttl` with a large sample size? I chose `volatile-lru` instead because it seemed safer. If we have many more `conj:` keys (3.4M) than `q:`, `as:` and `c:` keys (1M), it may be very likely that `volatile-ttl` could remove invalidation data by mistake and cause us to serve stale data?

If that happens, would we serve data just for a while, or indefinitely?

**Comment by mrmachine** [comment]
If `conj:` keys normally have an expiry (but I think this is disabled with the LRU setting and `volatile-lru` maxmemory policy), does that mean it is safe to just remove the whole key from time to time? Or does cacheops normally do some special handling before or when a `conj:` key expires?

**Comment by tumb1er** [comment]
Lots of questions.

Shortly, in docs both eviction strategies are considered as experimental, so they are just not suitable in your case. 

Worst case for inconsistent cache state is that you receive stale data, that was not invalidated because `conj:` key was evicted. Stale `q:` cache key _will_ expire in `timeout` seconds, and it depends on your application, is this a serious bug or not.

We've experienced similar problems, and developers found some ways to escape from ever-growing `conj:` keys. It was not silver bullet, just a bunch of fixes in certain cases.

**Comment by Suor** [comment]
> Does a the set from above include the as: keys for all @cached_as functions that include any queryset sample that filters against TABLENAME__slug=FOO?

Yes, it includes those.

> So Model1.objects.filter(slug=FOO) and Model2.objects.filter(model1__slug=FOO) in different @cached_as functions?

If both `Model1` and `Model2` refer to the same `TABLENAME` then yes, e.g. proxy models. If those models refer to the different tables then that will be separate conj sets.

> Why/how are all the keys in that set that I've tested returning (nil)?

Cache keys expire by time or max memory, their names are not removed from sets automatically.

> Are the keys themselves already expired, but the set has continued to accumulate more and more keys over time, never removing expired keys?

Yes. The set is only destroyed if matching changing event happens, i.e. `Model.objects.create(slug="FOO")`, or if that one was never updated within its lifetime.

So if you have constant reads, but no matching updates then cong sets will grow indefinitely. You might take a look at https://github.com/Suor/django-cacheops/pull/323 though, this is the script @tumb1er is mentioning.

> So the q: key for Article.objects.filter(is_active=True, category__slug=FOO) will go into conj:CATEGORY:slug=FOO AND conj:ARTICLE:is_active=true sets?

Yes, both, you can read on how cacheops works here http://hackflow.com/blog/2014/03/09/on-orm-cache-invalidation/.

> Could cacheops perhaps treat unique slug fields that never change differently?

cacheops currently ignores filters on `django_content_type` for reasons you describe. If you want to add your table to the list of ignored you'll need to change code [here](https://github.com/Suor/django-cacheops/blob/b59406a62c8265a9f3ace14c139e3f3be6c02407/cacheops/tree.py#L120). You can make a `CACHEOPS_*` setting for this and make a PR. This way we will be able to mainline it.

> Or remove expired keys from conj: sets periodically?

There exists [the conj-reaper command PR](https://github.com/Suor/django-cacheops/pull/323) I mentioned. If you give it a try and it works for you, then we'll work on merging that PR too. 

> Or, would we be better served by switching to volatile-ttl with a large sample size?

Neither of this options solve big conj sets issue. You will still need to user the conj-reaper. And if you do `max-memory volatile-lru` + `CACHEOPS_LRU = True` is superior.

> it may be very likely that volatile-ttl could remove invalidation data by mistake and cause us to serve stale data? If that happens, would we serve data just for a while, or indefinitely?

`volatile-ttl` is fundametally unsafe in this regard. You might get stale data with it. Each cache key still has TTL, so you can regulate maximum time you serve it.

> does that mean it is safe to just remove the whole key from time to time? Or does cacheops normally do some special handling before or when a conj: key expires?

It is not safe to remove conj key, this is why `max-memory` is dangerous, see above. Normally both cache keys and conj keys are expired by time, and since cache keys are guaranteed to have shorter life they expire before, so no orphan cache keys happening.

**TLDR.** Your best bet is `max-memory volatile-lru` + `CACHEOPS_LRU = True` + [conj-reaper](https://github.com/Suor/django-cacheops/pull/323) run periodically.

**Comment by Suor** [comment]
Please get back here if that works for you. Then we can look at what might be mainlined and what docs updated.

**Comment by mrmachine** [comment]
@Suor @tumb1er thanks both. I will try and conj-reaper and report back.

One more question...

> > So `Model1.objects.filter(slug=FOO)` and `Model2.objects.filter(model1__slug=FOO)` in different `@cached_as` functions?
> 
> If both `Model1` and `Model2` refer to the same `TABLENAME` then yes, e.g. proxy models. If those models refer to the different tables then that will be separate conj sets.

We're not using proxy models. And the biggest `conj:` set is a `Category` model which has 3 rows, with slugs that never change. We rarely query `Category` directly. We almost always only query other models, but filtered by category through the FK.

So now my question, if a different `conj` set is created for the model being queried, not the model being referenced in a filter lookup, how could we have gotten 670k cache keys for `Category`, a model with 3 rows that we rarely query directly?

To clarify, if I understand what you are saying correctly, `Article.objects.filter(category__slug=FOO)` would add a cache key to `conj:article:category__slug=FOO` and NOT `conj:category:slug=FOO`?

**Comment by Suor** [comment]
@mrmachine or I was inattentive, conj keys are built on model you filter by. So there is no such thing as `conj:article:category__slug=FOO`. 

BTW, now I remember, why I never hit this issue probably. We always shied away from joins like that and referenced categories by id everywhere, urls and so on. And when one absolutely needs to resolve a slug we used smth like:
```python
Model.objects.filter(category=Category.objects.get(slug="FOO"))
```

while using local cache for immutable things like that:
```python
CACHEOPS = {
    'app.Model': {'timeout': 60*15},
    'app.Category': {..., 'local_get': True},
}
```

This might ease it. However, if you are still going to use `max-memory` the recipe from TLDR is your best bet.

**Comment by mrmachine** [comment]
@Suor thanks for clarifying.

I'm testing a modified version of the conj-reaper script locally. Seems to be reducing the amount of data stored in the conj sets significantly (down to 1GB).

But it doesn't seem to have reduced the number of conj sets (still 3.4M), even though the reaper is reporting that it is removing ALL keys from many of these sets.

I thought Redis would automatically delete sets when its last member is removed?

Also, you mentioned using filter by ID instead of slug. How does that help?

Is it just that `Model.objects.filter(FOO__category_id=3)` vs `Model.objects.filter(FOO__category__slug=FOO)` will only be added to `conj:FOO:category_id=3` instead of `conj:CATEGORY:slug=FOO` AND `conj:FOO:category_id=3`?

And that cache keys in `conj:FOO:category_id` would be invalidated if any FOO is saved or deleted with `category_id=3`?

Also, when I reduce the min conj set size to 0 (to try and cut down that 1GB remaining conj set size), I see many many conj sets with a single expired key in them, named like `conj:FOO:id=465996`. Is it serving any purpose to have (seemingly) a single item conj set for every row of every table by PK?

**Comment by mrmachine** [comment]
The script It just finished running with min conj set size of 0. It took 50 minutes to complete. 2.9 million of the 3.4 million conj sets had only 1 cache key within.

So in addition to single item conj sets accounting for about 85% of the conj sets, they were probably accounting for about about 33% of the data size (~1GB) for all conj sets.

Are these single item conj sets required for normal invalidation of individual get queries for a single object, on save of that object?

**Comment by Suor** [comment]
> But it doesn't seem to have reduced the number of conj sets (still 3.4M), even though the reaper is reporting that it is removing ALL keys from many of these sets.
>
> I thought Redis would automatically delete sets when its last member is removed?

Redis does remove those. They might be populated immediately by some cache requests. Also is number of conj keys the issue?

> Also, you mentioned using filter by ID instead of slug. How does that help?

All those `Article.objects.filter(category_id=12)` queries won't be added to `conj:category:slug=FOO` so that won't grow. They will still be added to `conj:article:category_id=12`, but those will be at least separated by model and other query params, e.g. `Article.objects.filter(category_id=12, author_id=42)` will be added to `conj:article:author_id=42&category_id=12`, but not to `conj:article:category_id=12`. And they are added to some article conj keys now anyways.

`Model.objects.filter(FOO__category__slug=FOO)` is added to:
- `conj:model:`
- `conj:foo:`
- `conj:category:slug=FOO`

So it's at least one conj key for any model referenced.

> And that cache keys in conj:FOO:category_id would be invalidated if any FOO is saved or deleted with category_id=3?

Yes, this is why cacheops stores those sets, to del keys on events.

> Also, when I reduce the min conj set size to 0 (to try and cut down that 1GB remaining conj set size), I see many many conj sets with a single expired key in them, named like conj:FOO:id=465996. Is it serving any purpose to have (seemingly) a single item conj set for every row of every table by PK?

I did not write that reaper code, so there might be bugs there unknown to me. Also, that key might have expired after or while reaper was dealing with its conj key.

> Are these single item conj sets required for normal invalidation of individual get queries for a single object, on save of that object?

They are all required. However, having many conj sets usually means you are caching many things with different complex queries. It's common that those queries won't be hit ever, so cache is not helping with their performance. You might turn off cache for those, see README, Performance tips, 5.

**Comment by Suor** [comment]
Using `CACHEOPS_INSIDEOUT = True` is a blessed way to solve this now, see [Using memory limit](https://github.com/Suor/django-cacheops#contents)
